### ISSUE TITLE
division by 0 error in expression propagation

### ISSUE THREAD

**Issue by tly000** [cea-sec/miasm#1328]
Hi, 

I think the expression propagation is too lax when it comes to memory reads and writes. this can lead to some weird errors. in the following (simplified) code, the register "ECX" is replaced by 0 and this leads to a division by zero error:

```
MOV DWORD PTR [ESP + 0x14], 0x0
loop:
      ... more instructions
      MOV ECX, DWORD PTR [ESP + 0x14]
      INC ECX
      MOV DWORD PTR [ESP + 0x14], ECX
      ... more instructions
MOV ECX, DWORD PTR [ESP + 0x14]
TEST ECX, ECX
JNZ loc_2d027
RET
loc_2d027:
IDIV ECX
RET
```

this happens because the equivalence class for "ECX" contains `@32[ESP + 0x14]` and `0`. i can circumvent the error by changing the `may_interfer` function in data_flow.py to always return True if two memory expressions are involved.

here is the offending function:
```python
import pickle
import os
import sys

import graphviz

from miasm.core.locationdb import LocationDB
from miasm.ir.ir import AssignBlock
from miasm.expression.expression import ExprWalk
from miasm.analysis.simplifier import IRCFGSimplifierSSA
from miasm.analysis.machine import Machine

binary_size = 11274580
binary = bytearray(binary_size)
offset = 184016
instructions = [
    0x83,0xec,0x10,0x8b,0x44,0x24,0x18,0x53,0x55,0x56,0x57,0xc7,0x44,0x24,0x14,
    0x00,0x00,0x00,0x00,0xc7,0x44,0x24,0x18,0x00,0x00,0x00,0x00,0xc7,0x44,0x24,
    0x10,0x00,0x00,0x00,0x00,0x33,0xed,0x8b,0x4c,0x24,0x24,0x8a,0x54,0x0d,0x00,
    0x84,0xd2,0x75,0x08,0xc7,0x44,0x24,0x10,0x01,0x00,0x00,0x00,0x80,0xfa,0x7e,
    0x75,0x1e,0x8a,0x4c,0x0d,0x01,0x45,0x80,0xf9,0x6e,0x75,0x0a,0xc7,0x44,0x24,
    0x10,0x01,0x00,0x00,0x00,0xeb,0x0a,0x80,0xf9,0x7e,0x0f,0x95,0xc2,0x4a,0x83,
    0xe2,0x7e,0x84,0xd2,0x0f,0x84,0xd4,0x00,0x00,0x00,0x8b,0x4c,0x24,0x10,0x85,
    0xc9,0x0f,0x85,0xc8,0x00,0x00,0x00,0x80,0xfa,0x5f,0x88,0x54,0x24,0x28,0x75,
    0x06,0xb2,0x20,0x88,0x54,0x24,0x28,0x80,0xfa,0x61,0x72,0x12,0x80,0xfa,0x7a,
    0x77,0x0d,0x80,0x78,0x5b,0x61,0x73,0x07,0x80,0xc2,0xe0,0x88,0x54,0x24,0x28,
    0x33,0xc9,0x80,0xfa,0x96,0x76,0x54,0x8b,0x78,0x60,0x85,0xff,0x74,0x4d,0x8b,
    0x74,0x24,0x28,0x8b,0xde,0x81,0xe3,0xff,0x00,0x00,0x00,0x8a,0x9b,0x2a,0xf5,
    0x0a,0x00,0x80,0xfb,0xff,0x74,0x3a,0x81,0xe3,0xff,0x00,0x00,0x00,0xd1,0xe3,
    0x8a,0x93,0x78,0xf5,0x0a,0x00,0x88,0x54,0x24,0x1c,0x8a,0x93,0x79,0xf5,0x0a,
    0x00,0x80,0xfa,0xff,0x88,0x54,0x24,0x28,0x75,0x15,0x8b,0x4c,0x24,0x1c,0x81,
    0xe1,0xff,0x00,0x00,0x00,0xc1,0xe1,0x04,0x03,0xcf,0x32,0xd2,0x88,0x54,0x24,
    0x28,0x8b,0x74,0x24,0x28,0x8b,0x78,0x5c,0x85,0xff,0x74,0x21,0x8a,0x58,0x5a,
    0x3a,0xd3,0x72,0x1a,0x3a,0x50,0x5b,0x77,0x15,0x81,0xe6,0xff,0x00,0x00,0x00,
    0x81,0xe3,0xff,0x00,0x00,0x00,0x2b,0xf3,0xc1,0xe6,0x04,0x03,0xf7,0x8b,0xce,
    0x85,0xc9,0x74,0x17,0x0f,0xbf,0x49,0x0e,0x8b,0x54,0x24,0x18,0x03,0xd1,0x8b,
    0x4c,0x24,0x14,0x41,0x89,0x54,0x24,0x18,0x89,0x4c,0x24,0x14,0x8b,0x4c,0x24,
    0x10,0x45,0x85,0xc9,0x0f,0x84,0xe0,0xfe,0xff,0xff,0x8b,0x4c,0x24,0x14,0x85,
    0xc9,0x75,0x0a,0x33,0xc0,0x5f,0x5e,0x5d,0x5b,0x83,0xc4,0x10,0xc3,0x8b,0x44,
    0x24,0x18,0x5f,0x99,0x5e,0x5d,0xf7,0xf9,0x5b,0x83,0xc4,0x10,0xc3,0x8d,0xa4,
    0x24,0x00,0x00,0x00,0x00,0x8d,0x49,0x00,
]
binary[offset:offset + len(instructions)] = instructions

loc_db = LocationDB()
machine = Machine("x86_32")
mdis = machine.dis_engine(binary, loc_db=loc_db, follow_call=False)


class IRAModified(machine.ira):

    def get_out_regs(self, block):
        extra_regs = {self.arch.regs.EAX, self.arch.regs.EBX, self.arch.regs.ECX, self.arch.regs.EDX,
                      self.arch.regs.EBP, self.arch.regs.ESI, self.arch.regs.EDI, self.arch.regs.ESP,
                      self.arch.regs.float_st0}
        return super(IRAModified, self).get_out_regs(block).union(extra_regs)

    def call_effects(self, addr, instr):
        # removes stack pointer changes from call effects
        # TODO handle ST0 return and usercalls
        assignblks, extra = super(IRAModified, self).call_effects(addr, instr)
        out = []
        for assignblk in assignblks:
            dct = dict(assignblk)
            dct = {
                dst: src for (dst, src) in dct.items() if dst != self.sp
            }
            out.append(AssignBlock(dct, assignblk.instr))
        return out, extra


cfg = mdis.dis_multiblock(offset)
graphviz.Source(cfg.dot()).view("cfg")
ir_arch = IRAModified(loc_db)

# ircfg = pickle.load(open("ircfg.dat", 'rb'))
ircfg = ir_arch.new_ircfg_from_asmcfg(cfg)
assert len(ircfg.heads()) == 1

simplifier = IRCFGSimplifierSSA(Machine('x86_32').ira(ircfg.loc_db))
ircfg = simplifier.simplify(ircfg, ircfg.heads()[0])
graphviz.Source(ircfg.dot()).view("ircfg")
# assert ircfg != None
verify_esp_propagation(ircfg)
```

I found another function where the propagation of a `PUSH 0x01` instruction replaces the `IRDst = @32[ESP]` by `IRDst = 0x01`. I think there should be a simple way to just disable the symbolic memory expression propagation.

**Comment by serpilliere** [comment]
Ok, got it.
During the propagation, the algorithm try to find a fixed point, in which states are propagated, independently of the tests conditions. In one simplification, we try to apply constant propagation for the division, which tries to compute the division with a null denominator.
I made a PR https://github.com/cea-sec/miasm/pull/1329 to fix this.
Maybe you can try to apply it above the branch https://github.com/serpilliere/miasm/tree/fix_del_dummy_phi
It works here on the sample you gave
Can you confirm it @tly000  ?

By the way, thanks a lot for your precious feedbacks!

**Comment by tly000** [comment]
Thank you for the fast replies. This means the propagation would not actually have propagated the zero value, but crashed when trying. Have i understood that correctly?

**Comment by serpilliere** [comment]
Yep, exactly!

**Comment by tly000** [comment]
is there a simple way to disable the propagation of memory stores? im just concerned that the correctness could be broken if a memory location is modified by a called function or by indirect accesses.

**Comment by serpilliere** [comment]
Hi @tly000 
Your question is interesting, and you are correct: to be sound, we must delete memory aliases after call execution. For the moment, interferences are only computed between the propagated value and the *destinations* of every assignments it wants to cross.

To add this, you may add some test at: https://github.com/cea-sec/miasm/blob/b7c2b25f0d9d3a15a663dda8bd40f55d3bcd8a03/miasm/analysis/data_flow.py#L2067
At this location, the `classes` object contains the equivalence classes, and the `new_out` object contains the assignments we want to cross.
So basically, we may add some test here to see if at least one source in the rights values of `new_out` is a `call` operator, and if yes, delete every memory in the equivalence classes.

I will try to add this in a near future.
Thank you for the remark @tly000 .
I will add another issue, as it deserve a full entry for it :smile:

**Comment by tly000** [comment]
thanks!
